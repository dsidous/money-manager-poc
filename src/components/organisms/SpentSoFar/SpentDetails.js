import React, { Component } from 'react';
import { ScrollView, View, Text, Modal } from 'react-native';
import { connect } from 'react-redux';
import PieChart from 'react-native-pie-chart';
import AllSpentList from './AllSpentList';
import RegularSpentList from './RegularSpentList';
import DayToDaySpentList from './DayToDaySpentList';
import SpentHistory from './SpentHistory';


class SpentDetails extends Component {

  constructor() {
    super();

    this.state = {
      showModal: false,
    };
  }

  showHistory = (item) => {
    this.setState({ showModal: true });
  }

  hideHistory = () => {
    this.setState({ showModal: false });
  }

  render() {
    const data = this.props.spent[this.props.selected];
    const sliceColor = ['#F44336', '#2196F3', '#FFEB3B', '#4CAF50', '#FF9800'];
    const amounts = this.props.selected === 'allSpending'
      ? data.filter(item => item.title !== 'All').map(item => item.amount)
      : data.map(item => item.amount);
    let spendingList;

    if (this.props.selected === 'allSpending') {
      spendingList = <AllSpentList data={data} sliceColor={sliceColor} />;
    } else if (this.props.selected === 'regularSpending') {
      spendingList = <RegularSpentList data={data} sliceColor={sliceColor} showHistory={this.showHistory} />;
    } else {
      spendingList = <DayToDaySpentList data={data} sliceColor={sliceColor} />;
    }

    return (
      <View style={{ flex: 1 }}>
        <View style={{
          flex: 0.5,
          justifyContent: 'flex-start',
          alignContent: 'center',
          alignSelf: 'stretch',
          width: '100%',
        }}>
          <PieChart
            chart_wh={150}
            series={amounts}
            sliceColor={sliceColor}
            doughnut
            coverRadius={0.65}
            coverFill="#eee"
            style={{ alignSelf: 'center' }}
          />
        </View>
        <ScrollView
          contentContainerStyle={{ flexGrow: 1 }}
          style={{ flex: 4 }}
        >
          {spendingList}
        </ScrollView>

        <Modal
          animationType="slide"
          transparent={false}
          visible={this.state.showModal}
          onRequestClose={() => {
            Alert.alert('Modal has been closed.');
          }}>
          <SpentHistory hideHistory={this.hideHistory} />
        </Modal>
      </View>
    )
  };
};

const mapsStateToProps = (state) => {
  return {
    spent: state.spending.data,
  };
};

export default connect(mapsStateToProps)(SpentDetails);
