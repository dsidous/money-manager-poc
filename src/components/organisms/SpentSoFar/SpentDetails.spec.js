import React from 'react';
import { shallow, configure } from 'enzyme';
import configureStore from 'redux-mock-store';
import Adapter from 'enzyme-adapter-react-16';
import SpentDetails from './SpentDetails';

configure({ adapter: new Adapter() });

describe('spent details', () => {
  const initialState = {
    spending: {
      data: {
        allSpending: [
          {
            title: 'All',
            amount: 3000,
          },
        ],
        regularSpending: [
          {
            date: 'June 1',
            title: 'Camden Council Tax',
            amount: 120,
          },
          {
            date: 'June 2',
            title: 'British Gas',
            amount: 110,
          },
        ],
        dayToDaySpending: [
          {
            category: 'Groceries',
            amount: 200,
          },
        ],
      },
    },
  };

  const selected = 'regularSpending';
  const mockStore = configureStore();
  const store = mockStore(initialState);
  const wrapper = shallow(<SpentDetails selected={selected} store={store} />);

  it('renders as expected', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
