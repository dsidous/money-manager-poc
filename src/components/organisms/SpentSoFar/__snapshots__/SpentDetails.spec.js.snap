// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`spent details renders as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(SpentDetails)
    selected="regularSpending"
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "dispatch": [Function],
      "selected": "regularSpending",
      "spent": Object {
        "allSpending": Array [
          Object {
            "amount": 3000,
            "title": "All",
          },
        ],
        "dayToDaySpending": Array [
          Object {
            "amount": 200,
            "category": "Groceries",
          },
        ],
        "regularSpending": Array [
          Object {
            "amount": 120,
            "date": "June 1",
            "title": "Camden Council Tax",
          },
          Object {
            "amount": 110,
            "date": "June 2",
            "title": "British Gas",
          },
        ],
      },
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "storeSubscription": Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "dispatch": [Function],
        "selected": "regularSpending",
        "spent": Object {
          "allSpending": Array [
            Object {
              "amount": 3000,
              "title": "All",
            },
          ],
          "dayToDaySpending": Array [
            Object {
              "amount": 200,
              "category": "Groceries",
            },
          ],
          "regularSpending": Array [
            Object {
              "amount": 120,
              "date": "June 1",
              "title": "Camden Council Tax",
            },
            Object {
              "amount": 110,
              "date": "June 2",
              "title": "British Gas",
            },
          ],
        },
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "storeSubscription": Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": [Function],
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
